panos_upgrade_assurance/check_firewall.py

```diff
@@ -1194,3 +1201,173 @@ def run_snapshots(self, snapshots_config: Optional[List[Union[str, dict]]] = Non
             result[snap_type] = self._snapshot_method_mapping[snap_type]()
 
         return result
+
+    def run_health_checks(
+        self,
+        checks_configuration: Optional[List[Union[str, dict]]] = None,
+        report_style: bool = False,
+    ) -> Union[Dict[str, dict], Dict[str, str]]:
+        """Run device health checks.
+
+        This method provides a convenient way of running health check methods. For details on configuration see the
+        [health checks](/panos/docs/panos-upgrade-assurance/configuration-details#health-checks) documentation.
+
+        # Parameters
+
+        checks_configuration (list(str,dict), optional): (defaults to `None`) List of readiness checks to run.
+        report_style (bool): (defaults to `False`) Changes the output to more descriptive. Can be used when generating a report
+            from the checks.
+
+        # Raises
+
+        WrongDataTypeException: An exception is raised when the configuration is in a data type different then `str` or `dict`.
+
+        # Returns
+
+        dict: Results of all configured checks.
+
+        """
+        result = {}
+        checks_list = ConfigParser(
+            valid_elements=set(self._health_check_method_mapping.keys()),
+            requested_config=checks_configuration,
+        ).prepare_config()
+
+        for check in checks_list:
+            if isinstance(check, dict):
+                check_type, check_config = next(iter(check.items()))
+                if check_config is None:
+                    check_config = {}
+            elif isinstance(check, str):
+                check_type, check_config = check, {}
+            else:
+                raise exceptions.WrongDataTypeException(
+                    f"Wrong configuration format for check: {check}."
+                )  # NOTE checks are already validated in ConfigParser._extrac_element_name - this is never executed.
+
+            check_result = self._health_check_method_mapping[check_type](
+                **check_config
+            )  # (**) would pass dict config values as separate parameters to method.
+            result[check_type] = str(check_result) if report_style else {"state": bool(check_result), "reason": str(check_result)}
+
+        return result
+
+    def check_device_root_certificate_issue(self, fail_when_affected_version_only: bool = True) -> CheckResult:
+        """Checks whether the target device is affected by the Root Certificate Expiration issue;
+
+        https://live.paloaltonetworks.com/t5/customer-advisories/emergency-update-required-pan-os-root-and-default-certificate/ta-p/564672
+
+        This check will FAIL if so, allowing you to build upgrade logic based on when and how it's failed.
+
+        This check will fail in the following scenarios;
+            1. The device is running software that is affected by the issue AND is running out of date content
+                AND is NOT running the user-id service or data redistribution
```

Comment:
I think this check assumes some public cloud service is being used like (wildfire, dns security etc) otherwise it is not affected?

Reply:
Correct. The vast majority of customers would be. But either way, if they aren't, then they can elect not to use the test and it puts the onus back onto them. If this assumption turns out to be false, we can iterate and make it a bit more specific later on.